File: lib/Ntreev.Windows.Forms.Grid/EditingReason.cpp
//=====================================================================================================================
// Ntreev Grid for .Net 2.0.5190.32793
// https://github.com/NtreevSoft/GridControl
// 
// Released under the MIT License.
// 
// Copyright (c) 2010 Ntreev Soft co., Ltd.
// 
// Permission is hereby granted, free of charge, to any person obtaining a copy of this software and associated 
// documentation files (the "Software"), to deal in the Software without restriction, including without limitation the 
// rights to use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the Software, and to permit 
// persons to whom the Software is furnished to do so, subject to the following conditions:
// 
// The above copyright notice and this permission notice shall be included in all copies or substantial portions of the 
// Software.
// 
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE 
// WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR 
// COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR 
// OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
//=====================================================================================================================


#include "StdAfx.h"
#include "EditingReason.h"

namespace Ntreev { namespace Windows { namespace Forms { namespace Grid
{
    using namespace System::Drawing;
    using namespace System::Windows::Forms;

    EditingReason::EditingReason(Point pt)
        : data(pt), reasonType(EditingReasonType::Mouse)
    {

    }

    EditingReason::EditingReason(Keys key)
        : data(key), reasonType(EditingReasonType::Key)
    {

    }

    EditingReason::EditingReason(System::Char charKey)
        : data(charKey), reasonType(EditingReasonType::Char)
    {

    }

    EditingReason::EditingReason(int keyValue)
        : data(keyValue), reasonType(EditingReasonType::Ime)
    {

    }

    EditingReason::EditingReason(const GrEditingReason& reason)
    {
        switch(reason.editingType)
        {
        case GrEditingType_None:
            data = nullptr;
            break;
        case GrEditingType_Mouse:
            data = (Point)reason.location;
            break;
        case GrEditingType_Key:
            data = (Keys)reason.key;
            break;
        case GrEditingType_Char:
            data = (System::Char)reason.character;
            break;
        case GrEditingType_Ime:
            data = (int)reason.keyValue;
            break;
        }

        reasonType = (EditingReasonType)reason.editingType;
    }

    GrEditingReason EditingReason::ToNative()
    {
        GrEditingReason reason;

        switch(reasonType)
        {
        case EditingReasonType::Mouse:
            reason.location = this->Location;
            break;
        case EditingReasonType::Key:
            reason.key = (GrKeys)this->Key;
            break;
        case EditingReasonType::Char:
            reason.character = (wchar_t)this->Char;
            break;
        case EditingReasonType::Ime:
            reason.keyValue = this->KeyValue;
            break;
        }

        reason.editingType = (GrEditingType)reasonType;
        return reason;
    }

    EditingReasonType EditingReason::ReasonType::get()
    {
        return reasonType;
    }

    Point EditingReason::Location::get()
    {
        return (Point)data;
    }

    Keys EditingReason::Key::get()
    {
        return (Keys)data;
    }

    System::Char EditingReason::Char::get()
    {
        return (System::Char)data;
    }

    int EditingReason::KeyValue::get()
    {
        return (int)data;
    }

    System::Object^ EditingReason::Data::get()
    {
        return data;
    }
} /*namespace Grid*/ } /*namespace Forms*/ } /*namespace Windows*/ } /*namespace Ntreev*/

File: lib/WinformGridHost/EditingReason.cs
using Ntreev.Library.Grid;
using System;
using System.Collections.Generic;
using System.Drawing;
using System.Linq;
using System.Text;
using System.Windows.Forms;

namespace Ntreev.Windows.Forms.Grid
{
    public class EditingReason
    {
        private readonly EditingReasonType reasonType;
        private readonly object data;

        private EditingReason()
        {

        }

        public EditingReason(Point location)
        {
            this.reasonType = EditingReasonType.Mouse;
            this.data = location;
        }

        public EditingReason(Keys key)
        {
            this.reasonType = EditingReasonType.Key;
            this.data = key;
        }

        public EditingReason(char charKey)
        {
            this.reasonType = EditingReasonType.Char;
            this.data = charKey;
        }

        public EditingReason(int keyValue)
        {
            this.reasonType = EditingReasonType.Ime;
            this.data = keyValue;
        }

        internal EditingReason(GrEditingReason reason)
        {
            switch (reason.editingType)
            {
                case GrEditingType.None:
                    data = null;
                    break;
                case GrEditingType.Mouse:
                    data = (Point)reason.location;
                    break;
                case GrEditingType.Key:
                    data = (Keys)reason.key;
                    break;
                case GrEditingType.Char:
                    data = reason.character;
                    break;
                case GrEditingType.Ime:
                    data = (int)reason.keyValue;
                    break;
            }

            reasonType = (EditingReasonType)reason.editingType;
        }

        public EditingReasonType ReasonType
        {
            get { return this.reasonType; }
        }

        public Point Location
        {
            get { return (Point)this.data; }
        }

        public Keys Key
        {
            get { return (Keys)this.data; }
        }

        public char Char
        {
            get { return (char)this.data; }
        }

        public int KeyValue
        {
            get { return (int)this.data; }
        }

        public object Data
        {
            get { return this.data; }
        }

        public static readonly EditingReason None = new EditingReason();

        internal GrEditingReason ToNative()
        {
            GrEditingReason reason = new GrEditingReason();

            switch (reasonType)
            {
                case EditingReasonType.Mouse:
                    reason.location = this.Location;
                    break;
                case EditingReasonType.Key:
                    reason.key = (GrKeys)this.Key;
                    break;
                case EditingReasonType.Char:
                    reason.character = this.Char;
                    break;
                case EditingReasonType.Ime:
                    reason.keyValue = this.KeyValue;
                    break;
            }
            reason.editingType = (GrEditingType)this.reasonType;
            return reason;
        }
    }
}
